src/dune_engine/scheduler.ml

```diff
@@ -643,8 +676,101 @@ module Handler = struct
   type t = Config.t -> Event.t -> unit
 end
 
+module Timer : sig
+  type t
+
+  val create : Event.Queue.t -> frequency:float -> t
+
+  type sleeper
+
+  val await : sleeper -> [ `Finished | `Cancelled ] Fiber.t
+
+  val cancel : t -> sleeper -> unit Fiber.t
+
+  val sleep : t -> float -> sleeper
+
+  val close : t -> unit
+end = struct
+  type sleeper = [ `Finished | `Cancelled ] Fiber.Ivar.t
+
+  type t =
+    { events : Event.Queue.t
+    ; mutex : Mutex.t
+    ; frequency : float
+    ; mutable sleepers : (float * [ `Finished | `Cancelled ] Fiber.Ivar.t) list
+    ; mutable active : bool
+    }
+
+  let await = Fiber.Ivar.read
+
+  let cancel t sleeper =
+    Mutex.lock t.mutex;
+    t.sleepers <-
+      List.filter t.sleepers ~f:(fun (_, sleeper') -> not (sleeper' == sleeper));
+    Mutex.unlock t.mutex;
+    let* res = Fiber.Ivar.peek sleeper in
+    match res with
+    | None -> Fiber.Ivar.fill sleeper `Cancelled
+    | Some _ -> Fiber.return ()
+
+  let polling_loop t () =
+    let rec loop () =
+      Mutex.lock t.mutex;
```

Comment:
I find it strange that the pre-condition of "loop" is that the lock is not held, but the post-condition is that it's held. 
Wouldn't it be nicer if they coincided (e.g. both "held")?